code/addons/test/src/node/test-manager.ts

```diff
@@ -37,42 +39,69 @@ export class TestManager {
     this.vitestManager.startVitest().then(() => options.onReady?.());
   }
 
-  async restartVitest(watchMode = false) {
+  async restartVitest({ watchMode, coverage }: { watchMode: boolean; coverage: boolean }) {
     await this.vitestManager.vitest?.runningPromise;
     await this.vitestManager.closeVitest();
-    await this.vitestManager.startVitest(watchMode);
+    await this.vitestManager.startVitest({ watchMode, coverage });
   }
 
-  async handleConfigChange(payload: TestingModuleConfigChangePayload) {
-    // TODO do something with the config
-    const config = payload.config;
+  async handleConfigChange(
+    payload: TestingModuleConfigChangePayload<any, { coverage: boolean; a11y: boolean }>
+  ) {
+    if (payload.providerId !== TEST_PROVIDER_ID) {
+      return;
+    }
+    if (this.coverage !== payload.config.coverage) {
+      try {
+        this.coverage = payload.config.coverage;
+        await this.restartVitest({ watchMode: this.watchMode, coverage: this.coverage });
+      } catch (e) {
+        this.reportFatalError('Failed to change coverage mode', e);
+      }
+    }
   }
 
   async handleWatchModeRequest(payload: TestingModuleWatchModeRequestPayload) {
-    // TODO do something with the config
-    const config = payload.config;
-
     try {
       if (payload.providerId !== TEST_PROVIDER_ID) {
         return;
       }
 
       if (this.watchMode !== payload.watchMode) {
         this.watchMode = payload.watchMode;
-        await this.restartVitest(this.watchMode);
+        await this.restartVitest({ watchMode: this.watchMode, coverage: false });
       }
     } catch (e) {
       this.reportFatalError('Failed to change watch mode', e);
     }
   }
 
-  async handleRunRequest(payload: TestingModuleRunRequestPayload) {
+  async handleRunRequest(payload: TestingModuleRunRequestPayload<Config>) {
     try {
       if (payload.providerId !== TEST_PROVIDER_ID) {
         return;
       }
 
+      const allTestsRun = (payload.storyIds ?? []).length === 0;
+      if (this.coverage) {
```

Comment:
We only want to restart vitest, if `watchMode` is not already disabled, don't we?

Reply:
no unfortunately we always have to restart Vitest before coverage, because otherwise the coverage report's stale coverage, not taking file changes into account. I tried to make that clear from the comments below, I'll try to make them more explicit.

Reply:
is the comment better now?